### PR TITLE
fix(appset): add ApplicationSet ProgressiveSync handling to clean up old appStatus entries when Applications are removed or RollingSync is disabled

### LINKED ISSUE #13331: RollingSync in AppSets not removing deleted apps from status.applicationStatus
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

When using RollingSync in AppSets, I’ve noticed that the `status.applicationStatus` list does not remove apps that are no longer present, causing the status not to match the real value of the app. This behavior is not expected, and I believe it should always show the current status of the apps.

**To Reproduce**

1. Set up an AppSet with RollingSync enabled.
2. Add multiple apps to the AppSet.
3. Observe the `status.applicationStatus` list reflecting the correct status of the apps.
4. Delete one of the apps from the AppSet.
5. Observe that the deleted app is still present in the `status.applicationStatus` list and the status does not match the real value of the app.

**Expected behavior**

The `status.applicationStatus` list should always show the current status of the apps, and it should remove the apps that are no longer present in the AppSet.

**Screenshots**

<!-- If applicable, add screenshots to help explain your problem. `-->`
example

**Version**

```shell
argocd version
argocd: v2.6.7+5bcd846
  BuildDate: 2023-03-23T14:57:27Z
  GitCommit: 5bcd846fa16e4b19d8f477de7da50ec0aef320e5
  GitTreeState: clean
  GoVersion: go1.18.10
  Compiler: gc
  Platform: linux/amd64
```

**Logs**
As you can see, despite the fact that we deleted some of the apps, they are still listed in the endless list of `status.applicationStatus` as pending see below:

```
status:
  applicationStatus:
    - application: istiod
      lastTransitionTime: '2023-04-24T09:43:23Z'
      message: >-
        Application resource became Healthy, updating status from Progressing to
        Healthy.
      status: Healthy
      step: '1'
    - application: keycloak-realm-import
      lastTransitionTime: '2023-04-24T10:01:07Z'
      message: >-
        Application resource became Healthy, updating status from Progressing to
        Healthy.
      status: Healthy
      step: '3'
    - application: keycloak
      lastTransitionTime: '2023-04-24T13:54:57Z'
      message: >-
        Application moved to Pending status, watching for the Application
        resource to start Progressing.
      status: Pending
      step: '2'
    - application: external-dns
      lastTransitionTime: '2023-04-24T09:53:15Z'
      message: >-
        Application resource is already Healthy, updating status from Waiting to
        Healthy.
      status: Healthy
      step: '2'
    - application: istio-base
      lastTransitionTime: '2023-04-24T09:53:23Z'
      message: >-
        Application resource became Healthy, updating status from Progressing to
        Healthy.
      status: Healthy
      step: '1'
    - application: istio-ingress
      lastTransitionTime: '2023-04-24T09:53:23Z'
      message: >-
        Application resource is already Healthy, updating status from Waiting to
        Healthy.
      status: Healthy
      step: '2'
    - application: testkube
      lastTransitionTime: '2023-04-24T09:56:02Z'
      message: >-
        Application resource is already Healthy, updating status from Waiting to
        Healthy.
      status: Healthy
      step: '3'
    - application: cert-manager
      lastTransitionTime: '2023-04-24T09:54:16Z'
      message: >-
        Application resource became Healthy, updating status from Progressing to
        Healthy.
      status: Healthy
      step: '2'
    - application: nfs-server-provisioner
      lastTransitionTime: '2023-04-24T09:53:49Z'
      message: >-
        Application resource is already Healthy, updating status from Waiting to
        Healthy.
      status: Healthy
      step: '2'
    - application: istio-egress-sb
      lastTransitionTime: '2023-04-24T09:53:41Z'
      message: >-
        Application resource is already Healthy, updating status from Waiting to
        Healthy.
      status: Healthy
      step: '2'
    - application: jaeger
      lastTransitionTime: '2023-04-24T09:55:23Z'
      message: >-
        Application resource is already Healthy, updating status from Waiting to
        Healthy.
      status: Healthy
      step: '3'
    - application: strimzi-kafka-operator
      lastTransitionTime: '2023-04-24T09:54:10Z'
      message: >-
        Application resource became Healthy, updating status from Progressing to
        Healthy.
      status: Healthy
      step: '2'
    - application: vertical-pod-autoscaler
      lastTransitionTime: '2023-04-24T09:54:13Z'
      message: >-
        Application resource became Healthy, updating status from Progressing to
        Healthy.
      status: Healthy
      step: '2'
    - application: gitlab-runner-k8s
      lastTransitionTime: '2023-04-24T09:53:53Z'
      message: >-
        Application resource is already Healthy, updating status from Waiting to
        Healthy.
      status: Healthy
      step: '3'
    - application: kiali
      lastTransitionTime: '2023-04-24T13:50:30Z'
      message: Application has pending changes, setting status to Waiting.
      status: Waiting
      step: '1'
    - application: kube-hunter
      lastTransitionTime: '2023-04-24T09:53:23Z'
      message: >-
        Application resource is already Healthy, updating status from Waiting to
        Healthy.
      status: Healthy
      step: '3'
    - application: loki-promtail
      lastTransitionTime: '2023-04-24T09:53:56Z'
      message: >-
        Application resource is already Healthy, updating status from Waiting to
        Healthy.
      status: Healthy
      step: '3'
    - application: goldilocks
      lastTransitionTime: '2023-04-24T12:51:06Z'
      message: >-
        Application resource became Healthy, updating status from Progressing to
        Healthy.
      status: Healthy
      step: '2'
    - application: loki-distributed
      lastTransitionTime: '2023-04-24T09:54:33Z'
      message: >-
        Application resource is already Healthy, updating status from Waiting to
        Healthy.
      status: Healthy
      step: '3'
    - application: oauth2-proxy
      lastTransitionTime: '2023-04-24T13:50:34Z'
      message: Application has pending changes, setting status to Waiting.
      status: Waiting
      step: '3'
    - application: fluentd
      lastTransitionTime: '2023-04-24T12:48:19Z'
      message: >-
        Application resource became Healthy, updating status from Progressing to
        Healthy.
      status: Healthy
      step: '2'
  conditions:
    - lastTransitionTime: '2023-04-24T09:01:57Z'
      message: Successfully generated parameters for all Applications
      reason: ApplicationSetUpToDate
      status: 'False'
      type: ErrorOccurred
    - lastTransitionTime: '2023-04-24T14:05:39Z'
      message: ApplicationSet Rollout Rollout started
      reason: ErrorOccurred
      status: 'False'
      type: ParametersGenerated
    - lastTransitionTime: '2023-04-24T09:01:57Z'
      message: ApplicationSet up to date
      reason: ApplicationSetUpToDate
      status: 'True'
      type: ResourcesUpToDate
```

### PR DESCRIPTION
**Pull request by wmgroot**
Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)). 
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.

Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request.

Closes https://github.com/argoproj/argo-cd/issues/13331

### CODE REVIEW

**Comment by crenshaw-dev** [comment]
@wmgroot how far back should we cherry-pick this?

**Comment by wmgroot** [comment]
This bug is present in the initial release of Progressive Syncs, so it could be cherrypicked to the next patch release of 2.6. Do you think that's necessary for an alpha feature though?

**Comment by crenshaw-dev** [comment]
As long as it's safe, and the cherry pick isn't gnarly, I'm in favor of cherry-picking.

**Comment by wmgroot** [comment]
That sounds fine to me, I think the change should be clean. What's the process for a cherry-pick? Should I open a second PR against the `release-2.6` branch?

**Comment by crenshaw-dev** [comment]
We'll see how the bot copes 😄

File: applicationset/controllers/applicationset_controller.go [hunk 1]
```diff
@@ -148,19 +148,28 @@ func (r *ApplicationSetReconciler) Reconcile(ctx context.Context, req ctrl.Reque
...
+	if r.EnableProgressiveSyncs {
+		if applicationSetInfo.Spec.Strategy == nil && len(applicationSetInfo.Status.ApplicationStatus) > 0 {
+			log.Infof("Removing %v unnecessary AppStatus entries from ApplicationSet %v", len(applicationSetInfo.Status.ApplicationStatus), applicationSetInfo.Name)
 
-		for _, app := range applications {
-			appMap[app.Name] = app
-		}
+			err := r.setAppSetApplicationStatus(ctx, &applicationSetInfo, []argov1alpha1.ApplicationSetApplicationStatus{})
+			if err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to set AppSet application statuses for %v: %w", applicationSetInfo.Name, err)
```
Comment (ishitasequeira) [inline applicationset/controllers/applicationset_controller.go:157]: nit: `failed to clear previous AppSet application statuses`

**Review by ishitasequeira — APPROVED** [review]
I was able to test the PR with ProgressiveSyncs and was able to see the application statuses getting cleaned up. 

I also went through the code and didn't spot anything weird. Just a minor nit.

**Comment by wmgroot** [comment]
Updated and rebased.

**Review by crenshaw-dev — APPROVED** [review]

**Pull request merged**